### PR TITLE
feat: support query parameters

### LINKED ISSUE #24801: Support parameterized queries in `/api/v3/query_sql`
With recent work in core, should be able to make use of the [`iox_query_params`](https://github.com/influxdata/influxdb3_core/tree/main/iox_query_params) crate to support parameterized queries in the `/api/v3/query_sql` API.

### Design Considerations

* How to pass the parameter values? (e.g., URL parameters, in request body)
* How are parameters represented in query strings? (Datafusion supports numbered, e.g., `$1`, and named parameters, e.g., `$foo`, see [here](https://docs.rs/datafusion/latest/datafusion/common/enum.ParamValues.html))

### PR DESCRIPTION
**Pull request by hiltontj**
Closes #24801

### Summary

* This adds support for parameters in the `/api/v3/query_sql` and `/api/v3/query_influxql` APIs.
* The new parameter `params` is supported in the URL query string of a `GET` request, or in the JSON body of a `POST` request.
* Two new E2E tests were added to check successful GET/POST as well as error scenario when params are not provided for a query string that would expect them.

### Usage

Consider the following query containing the parameter `$bar`:
```sql
SELECT * FROM foo WHERE bar = $bar;
```

We can invoke this query via the `GET /api/v3/query_sql` API:
```http
GET /api/v3/query_sql?params=<params>[...]
```
where `<params>` is a stringified JSON object containing `name`:`value` pairs for provided parameters, e.g.,
```json
{
  "bar": "baz"
}
```

We can also invoke this query via the `POST /api/v3/query_sql` API:
```http
POST /api/v3/query_sql

{"params": {"bar": "baz"}, [...] }
```

`GET /api/v3/query_influxql` and `POST /api/v3/query_influxql` work the same way.

### Follow-on issues

* Support `params` in the v1 `/query` API (see #24805)
* Support `params` in `influxdb3_client` (see #24806)
* Support `params` in the `query` CLI (see #24807)
* Currently, if there are parameters in the query string that are not provided in `params`, a `500` error will be returned containing an informative error message (see the tests added in this PR). Ideally, a `400` would be returned (see #24811)

### CODE REVIEW

File: influxdb3/tests/server/query.rs [hunk 1]
```diff
@@ -270,6 +391,130 @@ async fn api_v3_query_influxql() {
...
+            .await
+            .unwrap();
+        let status = resp.status();
+        let body = resp.text().await.unwrap();
+
+        // TODO - it would be nice if this was a 4xx error, because this is really
+        //   a user error; however, the underlying error that occurs when Logical
+        //   planning is Datafusion::Internal, and is not so convenient to deal with.
+        //
+        //   I filed https://github.com/apache/arrow-datafusion/issues/9738 to see
```
Comment (mgattozzi) [inline influxdb3/tests/server/query.rs:506]: I feel for both of these comments it makes sense to remove the use of I as anyone who comes to read the code won't know who it is. I know it's you but they won't

File: influxdb3_server/src/http.rs [hunk 2]
```diff
@@ -477,6 +499,39 @@ where
...
+                    query_str: r.query_str,
+                    format: r.format,
+                    params: r.params.map(|s| serde_json::from_str(&s)).transpose()?,
+                }
+            }
+            Method::POST => {
+                let body = self.read_body(req).await?;
+                serde_json::from_slice(body.as_ref())?
+            }
+            _ => todo!(),
```
Comment (mgattozzi) [inline influxdb3_server/src/http.rs:524]: Did you mean to make this unreachable or did you mean to handle this branch?

File: influxdb3_server/src/http/v1.rs [hunk 3]
```diff
@@ -58,7 +58,8 @@ where
 
         let chunk_size = chunked.then(|| chunk_size.unwrap_or(DEFAULT_CHUNK_SIZE));
 
-        let stream = self.query_influxql_inner(database, &query).await?;
+        // TODO - params, currently passing None here:
```
Comment (mgattozzi) [inline influxdb3_server/src/http/v1.rs:61]: If this has an open issue mind putting it alongside the Todo?

**Review by mgattozzi — COMMENTED** [review]
Code wise I think this is fine. Mostly just had a few comments on comments and a possibly unhandled Todo!() Unless that was intentional

**Comment by erratic-pattern** [comment]
All of this looks reasonable to me. Is the query executor adding the parameters to the DataFusion logical plan at some point further down?

File: influxdb3_server/src/http.rs (hunk 2 above)
Comment (hiltontj) [reply influxdb3_server/src/http.rs:524]: Thanks for catching. I added the error variant for this then forgot to use it 🤦

File: influxdb3_server/src/http/v1.rs (hunk 3 above)
Comment (hiltontj) [reply influxdb3_server/src/http/v1.rs:61]: Good call-out. I opened https://github.com/influxdata/influxdb/issues/24805 and will update the comment.

**Comment by hiltontj** [comment]
> All of this looks reasonable to me. Is the query executor adding the parameters to the DataFusion logical plan at some point further down?

@erratic-pattern - the params are forwarded to the planner for SQL/InfluxQL here: https://github.com/influxdata/influxdb/blob/d34195045bac068b5152b92619ae2b228300a423/influxdb3_server/src/query_executor.rs#L121-L131

**Review by erratic-pattern — APPROVED** [review]

**Review by mgattozzi — APPROVED** [review]

**Review by pauldix — APPROVED** [review]
Great stuff, 🚢 !

**Pull request merged**